From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: zhoumingtao <zhoumingtao@loongson.com>
Date: Tue, 22 Feb 2022 07:48:39 +0000
Subject: boringssl add support for loongarch64

added loongarch64 build configuration

diff --git a/include/openssl/base.h b/include/openssl/base.h
index b630236641440dcb46ab43e074fe4fd338da7045..ec0740f4e1fc73d4ee80a6748fc4949cee59a03d 100644
--- a/include/openssl/base.h
+++ b/include/openssl/base.h
@@ -109,6 +109,9 @@ extern "C" {
 #define OPENSSL_64_BIT
 #elif defined(__riscv) && __SIZEOF_POINTER__ == 4
 #define OPENSSL_32_BIT
+#elif defined(__loongarch64)
+#define OPENSSL_64_BIT
+#define OPENSSL_LOONG64
 #elif defined(__pnacl__)
 #define OPENSSL_32_BIT
 #define OPENSSL_PNACL
