From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: zhoumingtao <zhoumingtao@loongson.com>
Date: Fri, 29 Apr 2022 03:27:12 +0000
Subject: dawn add support for loongarch64

added loongarch64 build configuration

diff --git a/src/dawn/common/Platform.h b/src/dawn/common/Platform.h
index 5e4f9d7ff4927a909f75d3888ee1451d174888fd..7987a05467e72a44b37265ee1edc8652a3bbfbdd 100644
--- a/src/dawn/common/Platform.h
+++ b/src/dawn/common/Platform.h
@@ -68,7 +68,7 @@
 #endif
 
 #if defined(_WIN64) || defined(__aarch64__) || defined(__x86_64__) || defined(__mips64__) || \
-    defined(__s390x__) || defined(__PPC64__)
+    defined(__s390x__) || defined(__PPC64__) || defined(__loongarch64)
 #    define DAWN_PLATFORM_64_BIT 1
 static_assert(sizeof(sizeof(char)) == 8, "Expect sizeof(size_t) == 8");
 #elif defined(_WIN32) || defined(__arm__) || defined(__i386__) || defined(__mips32__) || \
