From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mingtao Zhou <zhoumingtao@loongson.cn>
Date: Mon, 6 Jun 2022 07:56:50 +0000
Subject: swiftshader add support for loongarch64

using llvm 11

diff --git a/CMakeLists.txt b/CMakeLists.txt
index deb0c05aa8d412e4329c95749ea8aad33b00b5e0..e816ef3d319e696841095063efcd6d3e2d5350fa 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -60,6 +60,12 @@ elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "ppc*")
     else()
         message(FATAL_ERROR "Architecture is not supported")
     endif()
+elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "loongarch*")
+    if(CMAKE_SIZEOF_VOID_P EQUAL 8)
+        set(ARCH "loongarch64")
+    else()
+        set(ARCH "loongarch")
+    endif()
 else()
     if(CMAKE_SIZEOF_VOID_P EQUAL 8)
         set(ARCH "x86_64")
@@ -185,9 +191,15 @@ set(DEFAULT_REACTOR_BACKEND "LLVM")
 set(REACTOR_BACKEND ${DEFAULT_REACTOR_BACKEND} CACHE STRING "JIT compiler back-end used by Reactor")
 set_property(CACHE REACTOR_BACKEND PROPERTY STRINGS LLVM LLVM-Submodule Subzero)
 
-set(DEFAULT_SWIFTSHADER_LLVM_VERSION "10.0")
-set(SWIFTSHADER_LLVM_VERSION ${DEFAULT_SWIFTSHADER_LLVM_VERSION} CACHE STRING "LLVM version to use")
-set_property(CACHE SWIFTSHADER_LLVM_VERSION PROPERTY STRINGS "10.0")
+if(ARCH STREQUAL "loongarch64")
+    set(DEFAULT_SWIFTSHADER_LLVM_VERSION "11.0")
+    set(SWIFTSHADER_LLVM_VERSION ${DEFAULT_SWIFTSHADER_LLVM_VERSION} CACHE STRING "LLVM version to use")
+    set_property(CACHE SWIFTSHADER_LLVM_VERSION PROPERTY STRINGS "11.0")
+else()
+    set(DEFAULT_SWIFTSHADER_LLVM_VERSION "10.0")
+    set(SWIFTSHADER_LLVM_VERSION ${DEFAULT_SWIFTSHADER_LLVM_VERSION} CACHE STRING "LLVM version to use")
+    set_property(CACHE SWIFTSHADER_LLVM_VERSION PROPERTY STRINGS "10.0")
+endif()
 
 # If defined, overrides the default optimization level of the current reactor backend.
 # Set to one of the rr::Optimization::Level enum values.
@@ -507,7 +519,7 @@ else()
     endif()
     if(ARCH STREQUAL "mips64el")
         set_cpp_flag("-EL")
-        set_cpp_flag("-march=mips64r2")
+        set_cpp_flag("-march=loongson3a")
         set_cpp_flag("-mabi=64")
         set_cpp_flag("-fPIC")
         set_cpp_flag("-mxgot")
@@ -739,7 +751,7 @@ endif()
 if(WIN32)
     set(OS_LIBS odbc32 odbccp32 WS2_32 dxguid)
 elseif(LINUX)
-    set(OS_LIBS dl pthread)
+    set(OS_LIBS dl pthread stdc++fs)
     if(SWIFTSHADER_BUILD_WSI_WAYLAND)
         list(APPEND OS_LIBS "${WAYLAND}")
     endif(SWIFTSHADER_BUILD_WSI_WAYLAND)
diff --git a/src/Reactor/BUILD.gn b/src/Reactor/BUILD.gn
index 676f99f1a38ddf5c011021fc666c76f267b6a883..09f9fdf8b5ddf625a232f6409b1c7a3f070b2780 100644
--- a/src/Reactor/BUILD.gn
+++ b/src/Reactor/BUILD.gn
@@ -308,7 +308,11 @@ if (supports_subzero) {
 
 if (supports_llvm) {
   swiftshader_source_set("swiftshader_llvm_reactor") {
-    llvm_dir = "../../third_party/llvm-10.0"
+    if (current_cpu == "loong64") {
+      llvm_dir = "../../third_party/llvm-11.0"
+    } else {
+      llvm_dir = "../../third_party/llvm-10.0"
+    }
 
     deps = [
       ":swiftshader_reactor_base",
diff --git a/src/Reactor/ExecutableMemory.cpp b/src/Reactor/ExecutableMemory.cpp
index d35f3d5f76b96bb8fe6edc2caa45948abb3bc99d..356a7a24e18f154c447d803672b8f68e55207e09 100644
--- a/src/Reactor/ExecutableMemory.cpp
+++ b/src/Reactor/ExecutableMemory.cpp
@@ -158,6 +158,8 @@ static int memfd_create(const char *name, unsigned int flags)
 #			define __NR_memfd_create 356
 #		elif __x86_64__
 #			define __NR_memfd_create 319
+#       elif __loongarch64
+#           define __NR_memfd_create 279
 #		endif /* __NR_memfd_create__ */
 #		ifdef __NR_memfd_create
 	// In the event of no system call this returns -1 with errno set
diff --git a/src/Reactor/LLVMJIT.cpp b/src/Reactor/LLVMJIT.cpp
index b6ee7267da64869f5a1a9db0dd7db8f570671164..9ba396519bcf1f2089749b413df06bb74f998588 100644
--- a/src/Reactor/LLVMJIT.cpp
+++ b/src/Reactor/LLVMJIT.cpp
@@ -374,7 +374,7 @@ static uint32_t sync_fetch_and_op(uint32_t volatile *ptr, uint32_t val, F f)
 }
 #endif
 
-#if LLVM_VERSION_MAJOR >= 11 /* TODO(b/165000222): Unconditional after LLVM 11 upgrade */
+#if LLVM_VERSION_MAJOR > 11 /* TODO(b/165000222): Unconditional after LLVM 11 upgrade */
 class ExternalSymbolGenerator : public llvm::orc::DefinitionGenerator
 #else
 class ExternalSymbolGenerator : public llvm::orc::JITDylib::DefinitionGenerator
@@ -560,7 +560,7 @@ class ExternalSymbolGenerator : public llvm::orc::JITDylib::DefinitionGenerator
 	};
 
 	llvm::Error tryToGenerate(
-#if LLVM_VERSION_MAJOR >= 11 /* TODO(b/165000222): Unconditional after LLVM 11 upgrade */
+#if LLVM_VERSION_MAJOR > 11 /* TODO(b/165000222): Unconditional after LLVM 11 upgrade */
 	    llvm::orc::LookupState &state,
 #endif
 	    llvm::orc::LookupKind kind,
@@ -798,7 +798,7 @@ public:
 
 	~JITRoutine()
 	{
-#if LLVM_VERSION_MAJOR >= 11 /* TODO(b/165000222): Unconditional after LLVM 11 upgrade */
+#if LLVM_VERSION_MAJOR > 11 /* TODO(b/165000222): Unconditional after LLVM 11 upgrade */
 		if(auto err = session.endSession())
 		{
 			session.reportError(std::move(err));
diff --git a/src/Reactor/LLVMReactor.cpp b/src/Reactor/LLVMReactor.cpp
index 74a83a3ea1844b4b063e009e35145dedb8484420..e922642bc25ed1208a91fb3cf4e7ae5c21828798 100644
--- a/src/Reactor/LLVMReactor.cpp
+++ b/src/Reactor/LLVMReactor.cpp
@@ -383,7 +383,11 @@ std::string Caps::backendName()
 
 bool Caps::coroutinesSupported()
 {
+#if defined(__loongarch__)
+	return false;
+#else
 	return true;
+#endif
 }
 
 bool Caps::fmaIsFast()
@@ -1320,7 +1324,7 @@ Value *Nucleus::createAtomicAdd(Value *ptr, Value *value, std::memory_order memo
 {
 	RR_DEBUG_INFO_UPDATE_LOC();
 	return V(jit->builder->CreateAtomicRMW(llvm::AtomicRMWInst::Add, V(ptr), V(value),
-#if LLVM_VERSION_MAJOR >= 11
+#if LLVM_VERSION_MAJOR > 11
 	                                       llvm::MaybeAlign(),
 #endif
 	                                       atomicOrdering(true, memoryOrder)));
@@ -1330,7 +1334,7 @@ Value *Nucleus::createAtomicSub(Value *ptr, Value *value, std::memory_order memo
 {
 	RR_DEBUG_INFO_UPDATE_LOC();
 	return V(jit->builder->CreateAtomicRMW(llvm::AtomicRMWInst::Sub, V(ptr), V(value),
-#if LLVM_VERSION_MAJOR >= 11
+#if LLVM_VERSION_MAJOR > 11
 	                                       llvm::MaybeAlign(),
 #endif
 	                                       atomicOrdering(true, memoryOrder)));
@@ -1340,7 +1344,7 @@ Value *Nucleus::createAtomicAnd(Value *ptr, Value *value, std::memory_order memo
 {
 	RR_DEBUG_INFO_UPDATE_LOC();
 	return V(jit->builder->CreateAtomicRMW(llvm::AtomicRMWInst::And, V(ptr), V(value),
-#if LLVM_VERSION_MAJOR >= 11
+#if LLVM_VERSION_MAJOR > 11
 	                                       llvm::MaybeAlign(),
 #endif
 	                                       atomicOrdering(true, memoryOrder)));
@@ -1350,7 +1354,7 @@ Value *Nucleus::createAtomicOr(Value *ptr, Value *value, std::memory_order memor
 {
 	RR_DEBUG_INFO_UPDATE_LOC();
 	return V(jit->builder->CreateAtomicRMW(llvm::AtomicRMWInst::Or, V(ptr), V(value),
-#if LLVM_VERSION_MAJOR >= 11
+#if LLVM_VERSION_MAJOR > 11
 	                                       llvm::MaybeAlign(),
 #endif
 	                                       atomicOrdering(true, memoryOrder)));
@@ -1360,7 +1364,7 @@ Value *Nucleus::createAtomicXor(Value *ptr, Value *value, std::memory_order memo
 {
 	RR_DEBUG_INFO_UPDATE_LOC();
 	return V(jit->builder->CreateAtomicRMW(llvm::AtomicRMWInst::Xor, V(ptr), V(value),
-#if LLVM_VERSION_MAJOR >= 11
+#if LLVM_VERSION_MAJOR > 11
 	                                       llvm::MaybeAlign(),
 #endif
 	                                       atomicOrdering(true, memoryOrder)));
@@ -1370,7 +1374,7 @@ Value *Nucleus::createAtomicMin(Value *ptr, Value *value, std::memory_order memo
 {
 	RR_DEBUG_INFO_UPDATE_LOC();
 	return V(jit->builder->CreateAtomicRMW(llvm::AtomicRMWInst::Min, V(ptr), V(value),
-#if LLVM_VERSION_MAJOR >= 11
+#if LLVM_VERSION_MAJOR > 11
 	                                       llvm::MaybeAlign(),
 #endif
 	                                       atomicOrdering(true, memoryOrder)));
@@ -1380,7 +1384,7 @@ Value *Nucleus::createAtomicMax(Value *ptr, Value *value, std::memory_order memo
 {
 	RR_DEBUG_INFO_UPDATE_LOC();
 	return V(jit->builder->CreateAtomicRMW(llvm::AtomicRMWInst::Max, V(ptr), V(value),
-#if LLVM_VERSION_MAJOR >= 11
+#if LLVM_VERSION_MAJOR > 11
 	                                       llvm::MaybeAlign(),
 #endif
 	                                       atomicOrdering(true, memoryOrder)));
@@ -1390,7 +1394,7 @@ Value *Nucleus::createAtomicUMin(Value *ptr, Value *value, std::memory_order mem
 {
 	RR_DEBUG_INFO_UPDATE_LOC();
 	return V(jit->builder->CreateAtomicRMW(llvm::AtomicRMWInst::UMin, V(ptr), V(value),
-#if LLVM_VERSION_MAJOR >= 11
+#if LLVM_VERSION_MAJOR > 11
 	                                       llvm::MaybeAlign(),
 #endif
 	                                       atomicOrdering(true, memoryOrder)));
@@ -1400,7 +1404,7 @@ Value *Nucleus::createAtomicUMax(Value *ptr, Value *value, std::memory_order mem
 {
 	RR_DEBUG_INFO_UPDATE_LOC();
 	return V(jit->builder->CreateAtomicRMW(llvm::AtomicRMWInst::UMax, V(ptr), V(value),
-#if LLVM_VERSION_MAJOR >= 11
+#if LLVM_VERSION_MAJOR > 11
 	                                       llvm::MaybeAlign(),
 #endif
 	                                       atomicOrdering(true, memoryOrder)));
@@ -1410,7 +1414,7 @@ Value *Nucleus::createAtomicExchange(Value *ptr, Value *value, std::memory_order
 {
 	RR_DEBUG_INFO_UPDATE_LOC();
 	return V(jit->builder->CreateAtomicRMW(llvm::AtomicRMWInst::Xchg, V(ptr), V(value),
-#if LLVM_VERSION_MAJOR >= 11
+#if LLVM_VERSION_MAJOR > 11
 	                                       llvm::MaybeAlign(),
 #endif
 	                                       atomicOrdering(true, memoryOrder)));
@@ -1422,7 +1426,7 @@ Value *Nucleus::createAtomicCompareExchange(Value *ptr, Value *value, Value *com
 	// Note: AtomicCmpXchgInstruction returns a 2-member struct containing {result, success-flag}, not the result directly.
 	return V(jit->builder->CreateExtractValue(
 	    jit->builder->CreateAtomicCmpXchg(V(ptr), V(compare), V(value),
-#if LLVM_VERSION_MAJOR >= 11
+#if LLVM_VERSION_MAJOR > 11
 	                                      llvm::MaybeAlign(),
 #endif
 	                                      atomicOrdering(true, memoryOrderEqual),
diff --git a/src/Reactor/SubzeroReactor.cpp b/src/Reactor/SubzeroReactor.cpp
index 8fe7e31e2103d8444688cf412d828be2126b963f..95b0ae639e6cbdce04b76e022fb4e3c19e80dad7 100644
--- a/src/Reactor/SubzeroReactor.cpp
+++ b/src/Reactor/SubzeroReactor.cpp
@@ -316,6 +316,8 @@ private:
 		return false;
 #elif defined(__mips__)
 		return false;
+#elif defined(__loongarch__)
+		return false;
 #else
 #	error "Unknown architecture"
 #endif
@@ -664,6 +666,8 @@ std::vector<EntryPoint> loadImage(uint8_t *const elfImage, const std::vector<con
 	ASSERT(sizeof(void *) == 8 && elfHeader->e_machine == EM_AARCH64);
 #elif defined(__mips__)
 	ASSERT(sizeof(void *) == 4 && elfHeader->e_machine == EM_MIPS);
+#elif defined(__loongarch64)
+	ASSERT(sizeof(void *) == 8 && elfHeader->e_machine == EM_LOONGARCH);
 #else
 #	error "Unsupported platform"
 #endif
@@ -902,6 +906,8 @@ Nucleus::Nucleus()
 #elif defined(__mips__)
 	Flags.setTargetArch(Ice::Target_MIPS32);
 	Flags.setTargetInstructionSet(Ice::BaseInstructionSet);
+#elif defined(__loongarch64)
+    //TODO
 #else  // x86
 	Flags.setTargetArch(sizeof(void *) == 8 ? Ice::Target_X8664 : Ice::Target_X8632);
 	Flags.setTargetInstructionSet(CPUID::SSE4_1 ? Ice::X86InstructionSet_SSE4_1 : Ice::X86InstructionSet_SSE2);
diff --git a/src/Reactor/reactor.gni b/src/Reactor/reactor.gni
index 04fad6f8119aee6e005febb2b479cef3ce160607..37f638c32ae6482cb37e941324bd43f05a6d3183 100644
--- a/src/Reactor/reactor.gni
+++ b/src/Reactor/reactor.gni
@@ -10,7 +10,7 @@ import("//build_overrides/build.gni")
 
 declare_args() {
   # Subzero doesn't support ARM64, MIPS64, PPC64, and RISCV64 (only x86 and ARMv7a).
-  supports_subzero = current_cpu != "arm64" && current_cpu != "mips64el" && current_cpu != "ppc64" && current_cpu != "riscv64"
+  supports_subzero = current_cpu != "arm64" && current_cpu != "mips64el" && current_cpu != "ppc64" && current_cpu != "riscv64" && current_cpu != "loong64"
 }
 
 declare_args() {
diff --git a/src/System/Linux/MemFd.cpp b/src/System/Linux/MemFd.cpp
index 1319d1121279e422cbdbe007f0ed24cd338d2cd6..0faf3b317dcec7b7d753291389564e1b1d0fffc6 100644
--- a/src/System/Linux/MemFd.cpp
+++ b/src/System/Linux/MemFd.cpp
@@ -35,6 +35,8 @@
 #	define __NR_memfd_create 356
 #elif __x86_64__
 #	define __NR_memfd_create 319
+#elif __loongarch64
+#  define __NR_memfd_create 279
 #endif /* __NR_memfd_create__ */
 
 LinuxMemFd::~LinuxMemFd()
diff --git a/src/Vulkan/BUILD.gn b/src/Vulkan/BUILD.gn
index 1bb14a868f88dd7790a0fcca5bc07fd26ed557c7..2eafc747569f955b4a0cbf637a1db9049d0e9a42 100644
--- a/src/Vulkan/BUILD.gn
+++ b/src/Vulkan/BUILD.gn
@@ -180,6 +180,9 @@ swiftshader_shared_library("swiftshader_libvulkan") {
       "-Wl,-exported_symbols_list," +
           rebase_path("vk_swiftshader.exports", root_build_dir),
     ]
+    if (current_cpu != "x64") {
+      ldflags += [ "-Wl,-lz" ]
+    }
   } else if (is_linux || is_chromeos || is_fuchsia) {
     inputs = [
       "vk_swiftshader.lds",
